Objects/exceptions.c

```diff
@@ -632,6 +631,310 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+{
+    _PyObject_GC_UNTRACK(self);
+    ExceptionGroup_clear(self);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static int
+ExceptionGroup_traverse(PyExceptionGroupObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->msg);
+    Py_VISIT(self->excs);
+    return BaseException_traverse((PyBaseExceptionObject *)self, visit, arg);
+}
+
+static PyObject *
+ExceptionGroup_str(PyExceptionGroupObject *self)
+{
+    if (self->msg && PyUnicode_CheckExact(self->msg)) {
+        Py_INCREF(self->msg);
+        return self->msg;
+    }
+    else {
+        return BaseException_str((PyBaseExceptionObject *)self);
+    }
+}
+
+static PyObject* exceptiongroup_subset(PyExceptionGroupObject *orig,
+                                       PyObject *excs)
+{
+    /* Returns an ExceptionGroup with metadata from orig and exceptions from
+    the list excs.
+    excs is a subset of the exceptions in orig->excs (this is not checked).
+    */
+    PyObject *args = NULL;
+    PyObject *eg = NULL;
+    PyObject *tb = NULL;
+    PyObject *context = NULL;
+    PyObject *cause = NULL;
+
+    Py_ssize_t num_excs = PyList_GET_SIZE(excs);
+    if (num_excs < 0) {
+        return NULL;
+    }
+    else if (num_excs == 0) {
+        return Py_NewRef(Py_None);
+    }
+
+    args = PyTuple_Pack(2, Py_NewRef(orig->msg), Py_NewRef(excs));
+    if (!args) {
+        goto error;
+    }
+    eg = PyObject_CallObject(PyExc_ExceptionGroup, args);
+    if (!eg) {
+        goto error;
+    }
+    tb = PyException_GetTraceback((PyObject*)orig);
+    if (tb) {
+        int res = PyException_SetTraceback(eg, tb);
+        Py_XDECREF(tb);
+        if (res == -1) {
+            goto error;
+        }
+    }
+    context = PyException_GetContext((PyObject*)orig);
+    if (context) {
+        PyException_SetContext(eg, context);
+    }
+    cause = PyException_GetCause((PyObject*)orig);
+    if (cause) {
+        PyException_SetCause(eg, cause);
+    }
+    return eg;
+error:
+    Py_XDECREF(eg);
+    return NULL;
+}
+
+/* check whether exc matches the matcher. Matcher can be:
+
+* - a function, then exc matches if function(exc) returns true
+* - an exception type (or tuple thereof) then exc matches if
+*   PyErr_GivenExceptionMatches(exc, matcher)
+* - (LATER:) a collection of exception instances, then exc
+*   matches if it belongs to the collection (TODO: determine
+*   collection type and implement)
+*
+* TODO: create a matcher struct that has the PyObject + a code
+* specifying its type. Then we don't need the type checks here.
+*/
+static int exceptiongroup_project_check_match(PyObject *exc, PyObject *matcher)
+{
+    if (PyFunction_Check(matcher)) {
+        /* case 1: function matcher */
+        PyObject *exc_matches = PyObject_CallOneArg(matcher, exc);
+        return exc_matches != NULL && PyObject_IsTrue(exc_matches);
+    }
+    else {
+        /* case 2: exception type matcher */
+        return PyErr_GivenExceptionMatches(exc, matcher);
```

Comment:
how does this interact with with virtual subclasses? bpo url: https://bugs.python.org/issue12029

currently `except*` is described as:

> The matching of `except*` clauses against an `ExceptionGroup` is performed recursively, using the `ExceptionGroup.split()` method:

and the split method with a type is defined as:
> `eg.split(TypeError)`, is equivalent to `eg.split(lambda e: isinstance(e, TypeError))`.



so from the spec I'd expect that:
```
    def test_exception_match_calls_subclasscheck(self):
        class SubclassEvery(type):
            def __subclasscheck__(cls1, cls2):
                return not cls2 is ExceptionGroup

        class Special(Exception, metaclass=SubclassEvery):
            pass

        try:
            raise ValueError
        except *Special:  # caught
            pass

        try:
            raise ValueError
        except Special:  # not caught
            pass
```

but the code uses PyErr_GivenExceptionMatches which has the bpo-12029 behaviour

Reply:
`ExceptionGroup` cannot be subclassed, and for other exceptions this is exactly the same issue as bpo-12029. Or do you think it is a different issue here?

Reply:
In this case `eg.project(Special)`, is different from `eg.project(lambda e: isinstance(e, Special))`

Reply:
consider an isinstance_gem function that uses PyErr_GivenExceptionMatches:

```
>>> def isinstance_gem(instance, types):
...     return bool(ctypes.pythonapi.PyErr_GivenExceptionMatches(ctypes.py_object(instance), ctypes.py_object(types)))
... 
>>> isinstance_gem(ValueError(), Special)
False
>>> isinstance(ValueError(), Special)
True
```

Reply:
Due to the same (known) issue of bpo-12029, not a new issue introduced in this PEP. Is that correct?

Reply:
it's either an error in the PEP or an issue in this PR
the PEP reads:

> `eg.split(TypeError)`, is equivalent to `eg.split(lambda e: isinstance(e, TypeError))`.

if it's an error in the pep then it should instead be:

> ```
> def given_exception_matches(given, exc):
>     try:
>         raise given
>     except exc:
>         return True
>     except:
>         return False
> ```
> `eg.split(TypeError)`, is equivalent to `eg.split(lambda e: given_exception_matches(e, > TypeError))`.

otherwise should this code be changed to call `PyObject_IsInstance`?

Reply:
Ok, I understand thanks.